rj_gameplay/rj_gameplay/tactic/pass_tactic.py

```diff
@@ -1,261 +1,183 @@
+                self.assigned_roles.append(role(robot))
 
     def tick(
-        self, world_state: rc.WorldState, role_results: tactic.RoleResults
-    ) -> List[tactic.SkillEntry]:
+        self, world_state: stp.rc.WorldState
+    ) -> List[Tuple[int, RobotIntent]]:  # (id, intent)
         """
-        :return: A list of size 1 or 2 skills depending on which roles are filled and state of aiming
-        TODO: Come up with better timings for starting receive
+        FSM
+         - init: request passer
+         - on filled req: init_roles > tick passer's capture
+         - when passer ready to pass: request receiver
+         - on filled req: init_roles > tick passer pass
+         - on ball passed: tick receiver, release passer role
+         - when receiver done: done
         """
-        pivot_result = role_results[self.pivot_kick]
-        receive_result = role_results[self.receive]
-        if receive_result and receive_result[0].is_filled():
-            self.passer_cost.chosen_receiver = receive_result[0].role.robot
-            self.pivot_kick.skill.target_point = np.array(
-                receive_result[0].role.robot.pose[0:2]
-            )
-            self.passer_cost.potential_receiver = receive_result[0].role.robot
-            return [self.receive]
-        elif pivot_result and pivot_result[0].is_filled():
-            potential_receiver = self.find_potential_receiver(world_state)
-            self.pivot_kick.skill.target_point = np.array(
-                [potential_receiver.pose[0], potential_receiver.pose[1]]
-            )
-            return [self.pivot_kick]
-        return []
-
-    def is_done(self, world_state: rc.WorldState):
-        print(self.receive.skill.is_done(world_state))
-        return self.receive.skill.is_done(world_state)
+
+        role_intents = []
+
+        if self._state == "init":
+            # TODO: allow Plays to pass in this (and further below) cost fns,
+            #       otherwise behavior is not easy to manipulate
+            self._role_requests = [
+                (
+                    stp.role.cost.PickClosestRobot(world_state.ball.pos),
+                    passer.PasserRole,
+                )
+            ]
+            self._needs_assign = True
+
+            self._state = "init_passer_capture"
+
+        elif self._state == "init_passer_capture":
+            # assumes play has given new role_requests
+            self.init_roles(world_state)
+            self._state = "passer_capture"
+
+        elif self._state == "passer_capture":
+            # TODO: these lines are a little ugly, any fix?
+            passer_role = self.assigned_roles[0]
+            assert isinstance(passer_role, passer.PasserRole)
+            intent = passer_role.tick(world_state)
+
+            role_intents = [(passer_role.robot.id, intent)]
+
+            if passer_role.pass_ready:
+                self._state = "get_receiver"
+
+        elif self._state == "get_receiver":
```

Comment:
Is the passer pivoting in this step? I feel like the getting receiver step should be something that finishes right before the pass is executed.